src/plots/cartesian/axis_defaults.js

```diff
@@ -64,6 +65,10 @@ module.exports = function handleAxisDefaults(containerIn, containerOut, coerce,
         }
     }
 
+    containerOut._initialCategories = axType === 'category' ?
+        orderedCategories(letter, containerIn.categorymode, containerIn.categorylist, options.data) :
```

Comment:
We'll need to coerce `categorymode` and `cateogrylist` before sending them to `orderedCategories`.

Coercion is done in this scope by the [shortcut coerce function](https://github.com/monfera/plotly.js/blob/9c366ce959b604ca55c09ddf49b357e712fcf0eb/src/plots/cartesian/layout_defaults.js#L127) defined in `plots/cartesian/layout_defaults.js`. The shortcut coerce function is wrapper around [`Lib.coerce`](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/src/lib/coerce.js#L290-L312) which applies coercion rules based on the type of the user input. For example, an [enumerated attribute](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/src/lib/coerce.js#L34-L46) can only be set to the possible values listed in the [attributes definitions](https://github.com/monfera/plotly.js/pull/1/files#diff-0e41c2e162564438ff091d0ed6b5b455R451), if not, it gets coerced to the default value.

Moreover, we'll need to add logic around `categorymode` and `categorylist`. For instance, at the moment, `categorylist` only has an effect if `categorymode` is set to `array`. Meaning that if a user specifies `categorylist` without `categorymode`, the defaults categories will be shown even though it is clear that the user wanted custom categories. So, we can be smarter by overriding the `catogorymode` default if `categoylist` is an array, similar to what's being done currently for `tickmode` and `tickvals` in [`plots/cartesian/tick_value_defaults.js`](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/src/plots/cartesian/tick_value_defaults.js#L23-L27) (see tests [here](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/test/jasmine/tests/axes_test.js#L397-L490) for more info).

We may want to look up [`lib/nested_properly.js`](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/src/lib/nested_property.js) while familiarizing yourself with `Lib.coerce` (see its test cases [here](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/test/jasmine/tests/lib_test.js#L249-L436)).

Don't hesitate to ask me any questions about our coercion framework. Surprising, it may the trickiest part to understand about plotly.js.